### PR TITLE
feat(proguard): add support for frame signature

### PR DESCRIPTION
**Pull request by viglia**
The signature will be properly converted from bytecode types to java types, plus the method parameters will be used to map the frame in absence of line numbers.

### CODE REVIEW

File: crates/symbolicator-proguard/src/symbolication.rs [hunk 1]
```diff
@@ -241,6 +257,12 @@ impl ProguardService {
                                 mapped_frame.in_app = Some(true);
                             }
                         }
+                        // if there is a signature that has been deobfuscated,
+                        // add it to the mapped frame
+                        if deobfuscated_signature.is_some() {
+                            mapped_frame.signature =
+                                Some(deobfuscated_signature.as_ref().unwrap().format_signature());
+                        }
```
Comment (Swatinem) [inline crates/symbolicator-proguard/src/symbolication.rs:265]:
```suggestion
                        if let Some(signature) = &deobfuscated_signature {
                            mapped_frame.signature = Some(signature.format_signature());
                        }
```

File: crates/symbolicator-proguard/src/symbolication.rs [hunk 2]
```diff
@@ -184,19 +184,35 @@ impl ProguardService {
...
+                        return Some(deobfuscated_signature);
+                    }
+                }
+                None
+            });
+        let params = deobfuscated_signature
+            .as_ref()
+            .map_or(String::new(), |sig| {
+                String::from_iter(sig.parameters_types().map(|param| format!("{},", param)))
+            });
```
Comment (Swatinem) [inline crates/symbolicator-proguard/src/symbolication.rs:200]: Might be worth figuring out how to avoid allocating intermediate `String`s here.
I would also rather propagate the `None` here, instead of having an empty `String`.

Down below you can then rather match on `params`, as that is the thing you are using.

You are using the `deobfuscated_signature` here. I thought that the parameters in the proguard file are obfuscated as well? Or are those already deobfuscated?

This could really use a small integration test with a minimized proguard file.

File: crates/symbolicator-proguard/src/symbolication.rs [hunk 3]
```diff
@@ -184,19 +184,35 @@ impl ProguardService {
...
         let stack_frame = frame
             .lineno
             .map(|lineno| {
                 proguard::StackFrame::new(&frame.module, &frame.function, lineno as usize)
             })
             .or_else(|| {
-                frame.parameters.as_ref().map(|params| {
-                    proguard::StackFrame::with_parameters(
+                if deobfuscated_signature.as_ref().is_some() {
```
Comment (Swatinem) [inline crates/symbolicator-proguard/src/symbolication.rs:207]:
```suggestion
                if let Some(params) = &params {
```

File: crates/symbolicator-proguard/src/symbolication.rs (hunk 2 above)
Comment (viglia) [reply crates/symbolicator-proguard/src/symbolication.rs:200]: The parameters stored in the returned struct `deobfuscated_signature` are already converted to proper java types and deobfuscated (when necessary). Here we really just need to concatenate them.

File: crates/symbolicator-proguard/src/symbolication.rs [hunk 4]
```diff
@@ -184,17 +184,30 @@ impl ProguardService {
...
+        });
         let stack_frame = frame
             .lineno
             .map(|lineno| {
                 proguard::StackFrame::new(&frame.module, &frame.function, lineno as usize)
             })
             .or_else(|| {
-                frame.parameters.as_ref().map(|params| {
+                params.as_mut().map(|p| {
+                    p.pop();
```
Comment (Swatinem) [inline crates/symbolicator-proguard/src/symbolication.rs:206]: It took me quite some time to figure out that you are popping off the trailing `,` here. Please do that in the initial statement that creates `params`. It doesn’t need to be `mut` so you can mutate it 10 lines after creating it, where its very hard to follow whats actually happening.

Comment (viglia) [reply crates/symbolicator-proguard/src/symbolication.rs:206]: @Swatinem I've actually removed that already but I haven't pushed yet.

I'm cleaning the code. I'll ask for a new review once I'm done.

**Review by Swatinem — APPROVED** [review]
